todo-app/src/app/todo-list/todo-list.component.ts

```diff
@@ -0,0 +1,178 @@
+import { default as _rollupMoment } from 'moment';
+
+import {trigger, keyframes, animate, transition} from "@angular/animations";
+import * as keyfr from '../keyframes/keyframes';
+const moment = _rollupMoment || _moment;
+
+
+
+export const MY_FORMATS = {
+  parse: {
+    dateInput: 'LL',
+  },
+  display: {
+    dateInput: 'LL',
+    monthYearLabel: 'MMM YYYY',
+    dateA11yLabel: 'LL',
+    monthYearA11yLabel: 'MMMM YYYY',
+  },
+};
+
+@Component({
+  selector: 'app-todo-list',
+  templateUrl: './todo-list.component.html',
+  styleUrls: ['./todo-list.component.scss'],
+  encapsulation: ViewEncapsulation.None,
+  providers: [ TodoDataService,
+    {
+      provide: DateAdapter,
+      useClass: MomentDateAdapter,
+      deps: [MAT_DATE_LOCALE, MAT_MOMENT_DATE_ADAPTER_OPTIONS]
+    },
+
+    {provide: MAT_DATE_FORMATS, useValue: MY_FORMATS},
+  ],
+  animations: [
+    trigger( 'noteAnimation', [
+      transition('* => swipeLeft', animate(700, keyframes(keyfr.swipeLeft))),
+      transition('* => swipeRight', animate(700, keyframes(keyfr.swipeRight))),
+    ])
+  ]
+})
+
+export class TodoListComponent implements OnInit {
+
+  todos: ITodo[];
+  todo: ITodo = new Todo(  "", "", false, '');
+  form: FormGroup
+  list: FormGroup
+  searchInput: string
+  step = 0
+  animationState: string
+
+  constructor(private todoDataService: TodoDataService) { }
+
+  ngOnInit(): void {
+    this.form = new FormGroup({
+      title: new FormControl('', [Validators.required, Validators.minLength(2)]),
+      deadline: new FormControl(  '', Validators.required)
+    })
+    this.listTodo()
+  }
+
+  addTodo() {
+    let { title, deadline } = this.form.value;
+    deadline = moment(deadline).format('DD.MM.YYYY');
+    const  todo: ITodo = {
+      id: Date.now() + '',
+      title,
+      isDone: false,
+      deadline
+    }
+    this.todoDataService.addTodo(todo).subscribe( todo => {
+      this.todos.push(todo);
+      this.form.reset()
+    }, err => console.error(err))
+  }
+
+  listTodo() {
+    this.todoDataService.listTodo().subscribe((data: ITodo[])=>{
+      let list = [];
+      Object.keys(data).forEach(function(prop) {
+        let val = data[prop]
+        list.push(val)
+      })
+      this.todos = list;
+    });
+  }
+
+  sortByDone() {
+    this.todos.sort((a, b) => {
+      return <any>a.isDone - <any>b.isDone
+    })
+  }
+
+  mark( todo: Todo ) {
+    todo.isDone = !todo.isDone;
+    this.todoDataService.markTodo(todo).subscribe((data)=>{
+      this.sortByDone();
+      return todo;
+    });
+  }
+
+  delete( todo: Todo ) {
+    this.todoDataService
+      .deleteTodo(todo)
+      .subscribe(()=>{
+        this.todos = this.todos.filter(otherTodo => otherTodo.id !== todo.id)
+    }, err => console.error(err))
+  }
+
+  drop(event: CdkDragDrop<string[]>) {
+    moveItemInArray(this.todos, event.previousIndex, event.currentIndex);
+  }
+
+  sortDone() {
+    this.todoDataService.listTodo().subscribe((data: ITodo[])=>{
```

Comment:
I pretty sure that one 'this.todoDataService.listTodo()' is enough for whole component. We need to get data once and after that use existing data.

Reply:
Yes, I agree, but as we do this only once then after mark todo as Done the property will be changed in db and if we will not send second request to get the data we will get old todolist...

Reply:
Regarding listTodo(). Do you really need an additional request for sorting?

Reply:
Yes I do, otherwise sorting not going properly.

Reply:
It could be. You sort items in the app, not on the server. So we don't need additional request because we have necessary data.